### ISSUE TITLE
Thread p->IsMain() true

### ISSUE THREAD

**Issue by irmlab** [wxphp/wxphp#84]
I am translating the wxWindow wxThread sample from documentation. ("wxThread Class Reference)
The app crashes on Thread->Run() function. 
The thread does not enter in MyThread::Entry() function. The app crashes on exit, in all cases.

In the PHP debugger, I made some inspection on the created wxThread object:
With inspection on "<object>->IsMain()* the answer is ....  *true*   :-D))
And WindowsSeven 32 bits reports an Error in php5ts.dll.

    function  DoStartThread()
    {
        $this->m_pThread = new MyThread($this);
        
        if(0)  //added this bloc for debugging 
        {
                $ret= $this->m_pThread->Create(8000);  // Create() OK, but IsMain() is true !!
                if($ret != wxTHREAD_NO_ERROR)
                {
                    log_message("\nCan't create the thread!");
                    $this->m_pThread = null; return;
                    return;
                }
                // if($this->m_pThread->IsMain())   exit("\nThe new Thread IsMain() returns true");
        }

        $ret = $this->m_pThread->Run();       // The app will crash on function return....
        if($ret != wxTHREAD_NO_ERROR)
        {
            log_message("\nCan't run the thread!");
            $this->m_pThread = null;
            return;
        }
        // after the call to wxThread::Run(), the m_pThread pointer is "unsafe":
        // at any moment the thread may cease to exist (because it completes its work).
        //      To avoid dangling pointers OnThreadExit() will set m_pThread
        //      to NULL when the thread dies.
    }

====

BTW: what is the good way to use EVT_COMMAND and/or wxDECLARE_EVENT etc. to use with wxThreadEvent ?

The Documentation example uses :
   
    // declare a new type of event, to be used by our MyThread class:
    wxDECLARE_EVENT(wxEVT_COMMAND_MYTHREAD_COMPLETED, wxThreadEvent);
    wxDECLARE_EVENT(wxEVT_COMMAND_MYTHREAD_UPDATE, wxThreadEvent);
    ...
    wxDEFINE_EVENT(wxEVT_COMMAND_MYTHREAD_COMPLETED,wxThreadEvent)
    wxDEFINE_EVENT(wxEVT_COMMAND_MYTHREAD_UPDATE, wxThreadEvent)

What is the good PHP here ? To re-use some EVT_COMMAND already defined in the wxwidgets.php file ? For example, I wrote (so I can mirror the documentation) :
    
    define ("wxEVT_COMMAND_MYTHREAD_UPDATE", wxEVT_WIZARD_PAGE_CHANGING);
    define ("wxEVT_COMMAND_MYTHREAD_COMPLETED", wxEVT_WIZARD_FINISHED);

This should work, but I do not understand these macros, and I am not sure about the method.

Note: for dynamic event handling, ->Bind() is not implemented -- however I think the prototype is probably better for the wxPHP interface (->Connect() has multiple prototypes)

wxPHP is amazing, anyway, but maybe we can't have too much advanced programming.
Thanks !

**Comment by irmlab** [comment]
I did a basic working project with the implementation, so you can debug the DoStartThread()
(How do I send you a ZIP file ?)

**Comment by jgmdev** [comment]
While debugging with your wxThread sample port https://groups.google.com/forum/#!topic/wxphp/FdNbj_zy9H0 I did some modifications and got it working, mostly changed the Entry method:

```php
<?php
   /**
     * @return wxThread::ExitCode
     */
    function Entry()
    {
        while(!$this->TestDestroy())
        {
            wxMutexGuiEnter();
            // Causes a race condition or something related
            //log_error("\nPrinted from thread, CPU= " . wxThread::GetCPUCount());
            //This works
            print("\nPrinted from thread, CPU= " . wxThread::GetCPUCount());
            wxMutexGuiLeave();
            //wxQueueEvent($this->m_pHandler, new wxThreadEvent(wxEVT_COMMAND_MYTHREAD_UPDATE));
        }

        // signal the event handler that this thread is going to be destroyed
        // NOTE: here we assume that using the m_pHandler pointer is safe,
        //    (in this case this is assured by the MyFrame destructor)

        wxQueueEvent($this->m_pHandler, new wxThreadEvent(wxEVT_COMMAND_MYTHREAD_COMPLETED));
        return 0;     // 0== success (wxThread::ExitCode)
    }
```

The funny thing, it segfaults after first iteration of Entry, but if I run the code with **valgrind php testThreads.php** it works... So there seems to be some not funny race conditions happening which aren't popping around when running under valgrind.

Besides that, there also seems to be a issue with wxQueueEvent

**Comment by irmlab** [comment]
(Sorry for the late reply. )
No - As before, the modifications to wxEntry() cause a crash on $this->run.
Changing to print() is not a workaround.

BTW, the strangest thing is that wxPHP/Widgets says that the new Thread is the main thread !  

      ( In function DoStartThread($e) )

         $this->m_pThread->IsMain();   // debugger should NOT say "true"


I did not try to compile wxPHP on my machine, yet. As I started with wxPHP, the downloaded version is useful. And I switched to build some interfaces for my (very-old, command line) php scripts. 

wxPHP is a great tool, it allows complex interfaces, even with limited, basic widgets. 

Some strange things happens on Win7, like the wxListCtrl (it crashes) or some functions that are not recognized (but present in the wxPHP reference). But with some workarounds, the GUI interface works :) 

Thank a lot.
